packages/wp-now/src/pool.ts

```diff
@@ -0,0 +1,261 @@
+ */
+export class Pool {
+	instanceInfo = new Map(); // php => PoolInfo
+
+	spawner: () => Promise<any>; // Callback to create new instances.
+	maxRequests: number; // Max requests to feed each instance
+	maxJobs: number; // Max number of instances to maintain at once.
+
+	notifiers = new Map(); // Inverted promises to notify async code of backlogged item processed.
+	running = new Set(); // Set of busy PHP instances.
+	backlog = []; // Set of request callbacks waiting to be run.
+
+	constructor({
+		spawner = async (): Promise<any> => {}, 
+		maxRequests = 2000,
+		maxJobs = 5,
+	} = {}) {
+		this.spawner = spawner;
+		this.maxRequests = maxRequests;
+		this.maxJobs = maxJobs;
+		this[Reap]();
+		this[Spawn]();
+	}
+
+	/**
+	 * Find the next available idle instance.
+	 */
+	getIdleInstance() {
+		const sorted = [...this.instanceInfo].sort(
+			(a, b) => a[1].requests - b[1].requests
+		);
+
+		for (const [instance, info] of sorted) {
+			if (this.running.has(instance)) {
+				continue;
+			}
+
+			if (!info.active) {
+				continue;
+			}
+			return instance;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Queue up a callback that will make a request when an
+	 * instance becomes idle.
+	 */
+	async enqueue(item: (php: NodePHP) => Promise<any>) {
+		const idleInstance = this.getIdleInstance();
+
+		if (!idleInstance) {
+			// Defer the callback if we don't have an idle instance available.
+			this.backlog.push(item);
+
+			// Split a promise open so it can be accepted or
+			// rejected later when the item is processed.
+			const notifier = new Promise((accept, reject) =>
+				this.notifiers.set(item, {accept, reject})
+			);
+
+			// Return the notifier so async calling code
+			// can still respond correctly when the item
+			// is finally processed.
+			return notifier;
+		} else {
+			// If we've got an instance available, run the provided callback.
+
+			// Given an instance, create a new callback that will clean up
+			// after the instance processes a request, and optionally
+			// will also kick off the next request.
+			const onCompleted = (instance) => async () => {
+				this.running.delete(instance);
+
+				this[Reap]();
+				const newInstances = this[Spawn]();
+
+				// Break out here if the backlog is empty.
+				if (!this.backlog.length) {
+					return;
+				}
+
+				// This is the instance that completed
+				// so we can re-use it...
+				let nextInstance = instance;
+
+				// ... but, if we've just spanwed a fresh
+				// instance, use that one instead.
+				if (newInstances.size)
+					for (const instance of newInstances) {
+						nextInstance = instance;
+						break;
+					}
+
+				const next = this.backlog.shift();
+				const info = this.instanceInfo.get(nextInstance);
+
+				this.running.add(nextInstance);
+				info.requests++;
+
+				let request;
+
+				try {
+					// Don't ACTUALLY do anything until the
+					// instance is done spawning.
+					request = next(await nextInstance);
+				}
+				catch(error) {
```

Comment:
@dmsnell Let me know what you think of this logic. The alternative to a failed spawn is to return a 502 bad gateway.

Interested to know your thoughts.

Reply:
I'm a bit confused on what happens in these cases and probably need to think about it more. One thing that worries me is that if there is something intrinsic in the request that led to the failure, and we `unshift` it to the beginning of the queue, could we get into situations that lead to infinite and infinitely-fast re-death of newly spawned processes. I'm not sure if that worry is sound though, as I don't fully understand what's happening with these workers. It's not like they are OS processor or have the same characteristics of overhead and startup time.

In general, a `500` seems like an acceptable response. Why specifically a 502? The lines here are blurry.

Reply:
If I recall correctly an HTTP daemon will serve a 502 if it can't properly instantiate PHP to handle a request, but a 500 seems valid as well.

Reply:
@dmsnell @adamziel Lets revisit this, I don't want to forget about it.